### DISCUSSION TITLE
DF = 18 TC = 6 message bds30?

### DISCUSSION THREAD

**Discussion by ElSabio97** [junzis/pyModeS#106] [Q&A] [answered]
Hi again @junzis !

I'm downloading tiny bits of data using @xoolive `rawdata` query like this:

```
from traffic.data import eurofirs
from traffic.data import opensky
import pyModeS as pms

df = opensky.rawdata(
    start="2020-07-12 14:00",
    stop="2020-07-12 14:01",
    bounds = eurofirs['LECB'],
    cached = True
)

df.data['DF'] = df.data['rawmsg'].apply(lambda x: pms.df(x))
df.data['TC'] = df.data['rawmsg'].apply(lambda x: pms.typecode(x))

df.data['BITSlength'] = df.data['rawmsg'].apply(lambda x: len(pms.hex2bin(x))) #gives the ones that are 112 bits long, otherwise is30 will return error.

dfprebds30 = df.data.loc[df.data['BITSlength'] == 112].copy() #take only the ones 112 bits long to a new df.

dfprebds30['BDS30'] = dfprebds30['rawmsg'].apply(lambda x: pms.bds.bds30.is30(x)) #check which ones are actually BDS30

dfisBDS30 = dfprebds30.loc[dfprebds30['BDS30'] == True].copy() #take only the BDS30 to another df for easier data visualization.
```
If you run that code, you will get 77 messages that are BDS 30 BUT have DF18 and TC6.

From what I understand only DF 20/21 can provide BDS30 messages. What am I missing? 

If you want to check faster what I'm trying to say you can run the following commands to check one of those 77 messages:

```
print(pms.df("903433d7300007b48cb9a983db26"))
print(pms.typecode("903433d7300007b48cb9a983db26"))
print(pms.bds.bds30.is30("903433d7300007b48cb9a983db26"))
```

**Comment by junzis** [comment, marked as answer]
Indeed, you should always check the downlink format (make sure it is 20 or 21) before inferring the BDS code. 

I can't speak for the traffic library, but I think the rawdata may also have formats other than 20/21.

Also, don't use `bds30.is30()`, this only tells you if the message has the possibility of being a BDS 30 message. You should use `pms.bds.infer()` instead. This will give you all the possible BDS codes of the message. Then keep the ones with only BDS30.

**Reply by xoolive** [reply]
yes `rawdata` requests by default from all tables [(documentation)](https://traffic-viz.github.io/opensky_impala.html#requests-for-raw-data) but you may limit to DF 20 and 21 if you limit yourself to the `rollcall` table.
